design/baremetal-operator/early-cleaning.md

```diff
@@ -0,0 +1,128 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Early out-of-band cleaning
+
+This document proposes running driver-specific out-of-band clean
+steps.
+
+## Status
+
+implementable
+
+## Summary
+
+Depending on the selected driver, BMO will run zero or more out-of-band clean
+steps to ensure the node and the BMC are in the right state. Out-of-band clean
+steps are executed via BMC (e.g. using Redfish protocol) without booting the
+ironic agent (IPA).
+
+The first drivers to receive such handling will be `idrac` and
+`idrac-virtualmedia`. For them we will run the `management.known_good_state`
+clean step that resets iDRAC and clears its job queue.
+
+It is important that these actions run before we try to boot anything on the
+target node, be it the inspection ramdisk or a 3rd party installer ISO.
+
+## Motivation
+
+iDRAC will refuse to create a job if there is already a pending job of the same
+type. To make sure we start in a coherent state, we will reset iDRAC before
+doing anything else to it.
+
+### Goals
+
+* Provide a mechanism to run out-of-band clean steps before inspection.
+* Run `known_good_state` step for iDRAC.
+
+### Non-Goals
+
+* Run any in-band (i.e. using a ramdisk) clean steps.
+* Allow a user to provide the steps to run.
+
+## Proposal
+
+## Design Details
+
+All clean steps run asynchronously in the ironic API, even out-of-band steps
+that are reasonable fast (an order of seconds). When entering the `Registering`
+state, we will check if the driver supplies any reset clean steps. If yes, we
+will start manual cleaning and wait. Otherwise we'll move on.
+
+Resetting won't happen for externally provisioned host.
```

Comment:
Do we also want to reset the BMC when moving from externally provisioned to ready/inspecting? That would be a reason to add a separate Resetting state.

Reply:
Is it a real transition? I assumed externally provisioned things remain like that?

Reply:
Yes, there's a Spec flag for ExternallyProvisioned and we can't stop the user from clearing it (at least until we implement a validating webhook). We could just ignore it, but [currently](https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/host_state_machine.go#L271-L278) we do allow the host to transition to a normal state. If we didn't then users would have to delete the whole BMH to e.g. reprovision an externally provisioned host.

Reply:
I'll remove line 56.